app/player/cache.go

```diff
@@ -0,0 +1,199 @@
+// if `dir` does not exist, it will be created.
+// In other words, os.TempDir() should not be passed as a `dir`.
+func InitFSCache(opts *FSCacheOpts) (ChunkCache, error) {
+	storage, err := initFSStorage(opts.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Size == 0 {
+		opts.Size = defaultMaxCacheSize
+	}
+
+	r, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1e7, // number of keys to track frequency of (10M)
+		MaxCost:     opts.Size,
+		BufferItems: 64,
+		Metrics:     true,
+		OnEvict:     func(_, _ uint64, hash interface{}, _ int64) { storage.remove(hash) },
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &fsCache{storage, r}, nil
+}
+
+func initFSStorage(dir string) (*fsStorage, error) {
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, err
+	}
+
+	// Cache folder cleanup performed based on file names, chunk files will have a name of certain length.
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if dir == path {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return fmt.Errorf("subfolder %v found inside cache folder", path)
+		}
+		if len(info.Name()) != stream.BlobHashHexLength {
+			return fmt.Errorf("non-cache file found at path %v", path)
+		}
+		return os.Remove(path)
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return &fsStorage{dir}, nil
+}
+
+func (s fsStorage) remove(hash interface{}) {
+	var size int64
+	fi, _ := os.Stat(s.getPath(hash))
+	if fi != nil {
+		size = fi.Size()
+	}
+
+	if err := os.Remove(s.getPath(hash)); err != nil {
+		CacheLogger.Log().Errorf("failed to evict chunk file %v: %v", hash, err)
+		return
+	}
+	metrics.PlayerCacheSize.Sub(float64(size))
+	CacheLogger.Log().Infof("chunk file %v evicted", hash)
+}
+
+func (s fsStorage) getPath(hash interface{}) string {
+	return path.Join(s.path, hash.(string))
+}
+
+func (s fsStorage) open(hash interface{}) (*os.File, error) {
+	f, err := os.Open(s.getPath(hash))
+	if err != nil {
+		return nil, err
+	}
+	return f, nil
+}
+
+// Has returns true if cache contains the requested chink.
+// It is not guaranteed that actual file exists on the filesystem.
+func (c *fsCache) Has(hash string) bool {
+	_, ok := c.rCache.Get(hash)
+	return ok
+}
+
+// Get returns ReadableChunk if it can be retrieved from the cache by the requested hash
+// and a boolean representing whether chunk was found or not.
+func (c *fsCache) Get(hash string) (ReadableChunk, bool) {
+	if value, ok := c.rCache.Get(hash); ok {
+		f, err := c.storage.open(value)
+		if err != nil {
+			metrics.PlayerCacheErrorCount.Inc()
+			CacheLogger.Log().Errorf("chunk %v found in cache but couldn't open the file: %v", hash, err)
+			c.rCache.Del(value)
+			return nil, false
+		}
+		metrics.PlayerCacheHitCount.Inc()
+		cb, err := initCachedChunk(f)
+		f.Close()
+		if err != nil {
+			CacheLogger.Log().Errorf("chunk %v found in cache but couldn't read the file: %v", hash, err)
```

Comment:
how come we don't just roll up the error instead of returning true/false?

Reply:
It's better to handle this as a cache miss and log it for later analysis instead of bubbling up the error which would eventually result in a stream interruption.